src/EFCore.PG/Migrations/Internal/NpgsqlMigrationsAnnotationProvider.cs

```diff
@@ -55,6 +55,7 @@ public override IEnumerable<IAnnotation> For(IIndex index)
 
         public override IEnumerable<IAnnotation> For(IModel model)
             => model.GetAnnotations().Where(a =>
+                a.Name.StartsWith(RelationalAnnotationNames.DefaultSchema) ||
```

Comment:
I've looked into this a few times, and I could be missing something....but it just looks like their provider doesn't have to handle user-defined types that need to be stored as annotations.

In some ways, these annotations are quasi-dependent on each other, which feels like a problem that _ought_ to be encountered by other providers. But apparently not so far?

Reply:
Here's what's bothering me, and it could impact the general approach of this PR...

Including a DefaultSchema annotation on a *migration operation* (which is what it means for to be in `NpgsqlMigrationsAnnotationProvider`) implies that there has been a change in the default schema, and which needs to be somehow applied to the database via SQL. For example, if PostgreSQL had some of `SET default_schema` parameter which persists in the database, we would use an AlterModel migration operation to change that, and it would make sense to flow the DefaultSchema annotation on that operation.

Moreover, looking at the latest state of this PR, the DefaultSchema annotation doesn't seem to be used by NpgsqlMigrationsSqlGenerator... `GenerateEnumStatements()` iterates over all enum annotations on the AlterDatabaseOperation, which constructs PostgresEnum instances with fully-provided annotation name (already containing the schema) and value as strings. In other words, it seems that `BuildAnnotationName()` doesn't actually get called when generating migration SQL. So unless I'm missing something this can be removed?

One more concern I've stumbled upon when looking into this... When constructing the annotation, `BuildAnnotationName()` builds the annotation name with the schema - it tries to see if a schema was explicitly provided to the enum, defaults to the DefaultSchema annotation, and finally to null. So in the regular case (no schema, no DefaultSchema), we generate an annotation name such as `Npgsql:Enum:MyEnum`, no schema. Now, when the MigrationsSqlGenerator picks this annotation up, it calls the `Schema` property to get the enum's schema, but at that point you consult the DefaultSchema annotation again. This seem incorrect - if the enum annotation was *created* without a schema (i.e. the annotation name contains no schema), then it seems like we shouldn't override that when retrieving the schema later.

In other words, unless I'm missing something, I think it's safer and more correct to consult the DefaultSchema only at creation time. If this is right, we need to fix this in range as well.

Reply:
@roji Thanks for laying this all out. It was very helpful feedback. After some more thought, I agree that something isn't quite right with the current approach.

I think it stems from an early decision to have the `DefaultSchema` annotation eagerly bound into the enum annotation. That decision created the dependency between those annotations, the need for `DefaultSchema` to be passed to the migration operation, __and__ the inconsistencies in the annotation keys that I [mentioned](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/626#issuecomment-434518367) seeing in #626.

I've pushed two commits to remove the eager binding and then simplify the access patterns.

I'm including responses to your specific feedback below to elaborate on how this got tangled up, in case it's helpful for future work in this area:
___

> Here's what's bothering me...
>
> Including a DefaultSchema annotation on a _migration operation_ (which is what it means for to be in `NpgsqlMigrationsAnnotationProvider`) implies that there has been a change in the default schema, and which needs to be somehow applied to the database via SQL.

Right, that bothered me too at first, but I justified it with the assumption that we needed to allow for late-bound schema changes. But that's not right.

The mistake was binding one annotation to another in the first place. The default schema is describing to the _migration model_ where types should be placed if no schema is specified, so it shouldn't be consulted until the SQL is being generated for the migration. 

> Moreover, looking at the latest state of this PR, the DefaultSchema annotation doesn't seem to be used by NpgsqlMigrationsSqlGenerator... `GenerateEnumStatements()` iterates over all enum annotations on the AlterDatabaseOperation, which constructs PostgresEnum instances with fully-provided annotation name (already containing the schema) and value as strings.

That was the hiccup. Because of the way that the annotations were constructed with `schema.name` as the key, the annotation name may or may not have contained a schema. This was inconsistent when the source model had a `DefaultSchema` annotation, with the order of configuration playing into it.

The initial approach here was to "fix" this by doubling-down and yielding the `DefaultSchema` annotation to the `AlterDatabaseOperation`. But the answer is actually the exact opposite: _we should __never__ bind the value of one annotation into another_.

If no schema is given explicitly, then it's left null until generation-time when the default schema can be consulted from the appropriate model, rather than from the operation.

> In other words, unless I'm missing something, I think it's safer and more correct to consult the DefaultSchema only at creation time. If this is right, we need to fix this in range as well.

I think we agree here, assuming that when you say "at creation time", you mean when `Generate{Create,Drop}Enum` is called, rather than when the annotation is constructed.

Reply:
I actually think we may be mis-communicating over this :)

> I think it stems from an early decision to have the DefaultSchema annotation eagerly bound into the enum annotation.

I think you actually had it right - see below.

> I think we agree here, assuming that when you say "at creation time", you mean when Generate{Create,Drop}Enum is called, rather than when the annotation is constructed.

By "creation time", I actually did mean annotation creation time, and not SQL generation time.

Let me explain. First, let's look at what happens when a migration is generated for regular table creation. For a model that has `.HasDefaultSchema("some_schema")`, EF Core will generate a create table migrations follows:

```c#
migrationBuilder.CreateTable(
    name: "some_table",
    schema: "some_schema",
    columns: ...);
```

In other words, the DefaultSchema is baked into the migration when it is generated. We don't get a migration with a null schema, deferring the "binding" with the default schema until SQL generation time - the schema is eagerly inserted into the generated migration. I think enum/range types should get the exact same treatment: the only difference is that unlike with tables, the schema (like everything else) is managed via annotations, but that shouldn't really make a difference.

Unless I'm mistaken, the idea here is that someone may add a migration creating a table, and then change the DefaultSchema on the model. If DefaultSchema is bound late, at SQL generation time, that means that the table is created with the new DefaultSchema, even though the DefaultSchema was modified after the table creation migration was created - probably not the desired behavior. I think the goal is that once a migration is created, it ideally has all the information needed to do everything, within itself, without any need to consult anything else. This doesn't work 100% in all scenarios, but we should definitely try to keep it that way as much as possible.

I admit it's a bit of a confusing situation, and I'm still not sure I have the mental model completely right about everything. If you still feel unsure about any of this we can reach out to one of the EF Core team members for verification.

Reply:
It sounds like we are on the same page for the design _goals_ (i.e. self-contained migrations). I'm just not sure how to _implement_ it with the current architecture...

> By "creation time", I actually did mean annotation creation time, and not SQL generation time.

> ```c#
> migrationBuilder.CreateTable(
>    name: "some_table",
>    schema: "some_schema",
>    columns: ...);
> ```
> In other words, the DefaultSchema is baked into the migration when it is generated. We don't get a migration with a null schema, deferring the "binding" with the default schema until SQL generation time - the schema is eagerly inserted into the generated migration. 

> I think enum/range types should get the exact same treatment: the only difference is that unlike with tables, the schema (like everything else) is managed via annotations, but that shouldn't really make a difference.

It really seems like management via annotations _does_ make a difference in this case.

The trouble I'm having with this is that when the default schema is eagerly bound into the annotation name, there's no recourse if a default schema is added (or updated) later in model configuration. For example:

```c#
NpgsqlConnection.GlobalTypeMapper.MapEnum<SomeEnum>("test.some_enum");

modelBuilder
    // The annotation name is constructed as `Npgsql:Enum:some_enum`
    .ForNpgsqlHasEnum<SomeEnum>()
    // But after this, it *ought* to be `Npgsql:Enum:test.some_enum`
    .HasDefaultSchema("test");

// 42704: type "test.some_enum" does not exist
```

In this case, late-binding solves the problem, albeit imperfectly.

_TBH I would rather see the default schema only ever used at generation-time so that I could move many tables/functions/types to a new schema by changing the default schema. But that's an opinion for another time..._

> I think the goal is that once a migration is created, it ideally has all the information needed to do everything, within itself, without any need to consult anything else. This doesn't work 100% in all scenarios, but we should definitely try to keep it that way as much as possible.

I agree that this should be our goal. 

Perhaps we're looking at this too late in the process. Would it be crazy to consider replace [`MigrationsModelDiffer`](https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs)? 

It's currently in `Microsoft.EntityFrameworkCore.Migrations.Internal`, but it seems like it might give us the hooks to properly handle these as self-contained migration operations.

Reply:
> I think that rather than override `MigrationsModelDiffer`, this could be done cleanly in `NpgsqlMigrationsAnnotationProvider`, no? It's responsible for receiving annotations from the model and optionally providing annotations to the migrations process. That seems like a good spot to bake the default schema in, what do you think?

Oh, so rather than yield back the `DefaultSchema`, we would reconstruct the enum (and range) annotations and then yield back the "properly baked" version?

I hadn't considered that, but it sounds like it would work (and beats overriding a public-internal class). I'll give it a try this morning and report back.

Reply:
@ajcvickers you may be interested in this conversation on default schema and when it's "bound" into other annotation which depend on it (e.g. annotation which defines an enum type on the model).

Reply:
@AndriySvyryd @bricelam may have some thoughts on this.